File: python/src/tools/projects/get_projects.py
import json
from pydantic import BaseModel
from src.tools.types import Context, Tool, ToolResult, TextContent


class GetProjectsParams(BaseModel):
    pass


async def get_projects_handler(context: Context, _params: GetProjectsParams) -> ToolResult:
    org_id = await context.get_org_id()
    projects_result = await context.api.organizations().projects(org_id).list()

    if not projects_result.success:
        raise Exception(f"Failed to get projects: {projects_result.error}")

    return ToolResult(content=[TextContent(text=json.dumps([project.model_dump() for project in projects_result.data]))])


def get_projects_tool() -> Tool[GetProjectsParams]:
    return Tool(
        name="projects-get",
        description="""
        - Fetches projects that the user has access to - the orgId is optional. 
        - Use this tool before you use any other tools (besides organization-* and docs-search) to allow user to select the project they want to use for subsequent requests.
        """,
        schema=GetProjectsParams,
        handler=get_projects_handler
    )

File: typescript/src/tools/projects/getProjects.ts
import { ProjectGetAllSchema } from "@/schema/tool-inputs";
import { getToolDefinition } from "@/tools/toolDefinitions";
import type { Context, Tool } from "@/tools/types";
import type { z } from "zod";

const schema = ProjectGetAllSchema;

type Params = z.infer<typeof schema>;

export const getProjectsHandler = async (context: Context, _params: Params) => {
	const orgId = await context.stateManager.getOrgID();

	if (!orgId) {
		throw new Error("API key does not have access to any organizations.");
	}

	const projectsResult = await context.api.organizations().projects({ orgId }).list();

	if (!projectsResult.success) {
		throw new Error(`Failed to get projects: ${projectsResult.error.message}`);
	}

	return {
		content: [{ type: "text", text: JSON.stringify(projectsResult.data) }],
	};
};

const definition = getToolDefinition("projects-get");

const tool = (): Tool<typeof schema> => ({
	name: "projects-get",
	title: definition.title,
	description: definition.description,
	schema,
	handler: getProjectsHandler,
	annotations: {
		destructiveHint: false,
		idempotentHint: true,
		openWorldHint: true,
		readOnlyHint: true,
	},
});

export default tool;
